### PR TITLE
Zpool iostat show ssd

### PR DESCRIPTION
**Pull request by inkdot7**
Some questions:
- Currently the output is narrow, only using one character and the header overlaps the column separating spaces:
  zpool_iostat_ssd
  Should more space be used, and the info spelled out (yes, no and mix)?
- Since it adds a field and thus may break scripts, should it depend on a command-line option to zpool?
- The question in the second commit for module/zfs/vdev.c

### CODE REVIEW

**Comment by inkdot7** [comment]
@rlaager thanks for the suggestion. I have changed in a temporary branch zpool_iostat_show_ssd_2 of my zfs clone.  Do I just forcefully push to the branch for the pull request?  Wondering, as I guess that will kill some history the suggestion is based on.  (Although this is a rather harmless one.)

**Comment by rlaager** [comment]
I believe a force push will work. It's fine to throw away the history. That's normal in pull requests.

**Comment by kernelOfTruth** [comment]
@inkdot7 force-pushing is fine, I and the others do the same :)

**Comment by inkdot7** [comment]
@rlaager new version pushed.

**Comment by inkdot7** [comment]
Thanks @rlaager for investigating this!
1) The reason it was put in `zpool iostat` is that I find it very useful there when developing #4365 (which btw has reached an operational state, perhaps warranting some discussion too :-) .  These commits are an non-required part of that development).
4) Advantage is that one does not have to remember/cross-check which device is what kind (did the SDD become sda or sdb during boot today...?), the interesting ones are gotten directly.
2) Being Linux-specific could be another reason to hide it unless `zpool` is given an option flag?
3) Well... would that be a reason to be able to see what ZFS thinks (as told by the kernel)?
I would also be happy with it in `zpool status`, but found most use in `iostat` so far.

**Comment by inkdot7** [comment]
Also `zpool iostat` shows the individual vdev capacities/allocations, making this a useful location.  (And I thought it would disturb less here than in `status`.)

**Comment by Rudd-O** [comment]
Seconding the "4) Why not just use standard OS tools (e.g. hdparm) and interfaces (e.g. /proc) to determine this? Why does this need to exist in the ZFS tools at all?".

**Comment by rlaager** [comment]
@behlendorf, I recommend rejecting this pull request, primarily on the basis that it duplicates standard OS tool functionality without sufficient reason.

**Comment by inkdot7** [comment]
Well, whatever happens to this PR, I'll keep the branch around.  It could be that once dedicated vdev metadata / small block allocation functionality is included, it is found to be good to give the user a convenient way too see that the vdevs configured are of the intended nonrotational kinds.

**Comment by rlaager** [comment]
If you end up merging it with that work, that's another argument in favor of showing this in `zpool status` instead of `zpool iostat`. That patch shows the metadata types in `zpool status`.

**Comment by inkdot7** [comment]
The allocation class version shows them in both `status` and `iostat -v`.

**Comment by behlendorf** [comment]
> 1) This probably makes more sense in zpool status. The rotational state of disks is not changing second-by-second.

In the past I've wanted this exact functionality in both commands.  It's helpful when you want to see at a glance if a device is performing roughly within expectations.  For example, I don't expect many IOPs out of a HDD but I do from an SSD.

> 2) This is, in practice, Linux-specific, which may argue against its acceptance at all.

That's alright, it just needs to be a optional.  We shouldn't have to limit the Linux implementation due to the lack of functionality on other platforms.

> 3) The kernel's idea of what is non-rotational may not be accurate.

This is exactly the reason why this should be exposed.  Internally ZFS uses this value as one parameter in controlling how/when/if requests are merged.  Knowing what the kernel actually thinks about this disk can be insightful.

> 4) Why not just use standard OS tools (e.g. hdparm) and interfaces (e.g. /proc) to determine this? Why does this need to exist in the ZFS tools at all?

Because often it's laborious to map the vdev name back to the device to determine this information.  We didn't need to add the `-gLP` options either but they're convenient, d2f3e292dccab23e47ade3c67677a10f353b9e85.

That all being said I'm not a huge fan of the current output.  Can we do something more subtle like leave the default output as is, and optionally add a `*` (or other symbol) after the vdev name to indicate it's non-rotational.  Better suggestions welcome!

```
$ sudo zpool iostat -Pvs rpool
                capacity     operations    bandwidth
pool         alloc   free   read  write   read  write
-----------  -----  -----  -----  -----  -----  -----
rpool*       9.74G  20.0G      0      1  1.32K  15.0K
  /dev/sdd1* 9.74G  20.0G      0      1  1.32K  15.0K

$ sudo zpool status -Ps rpool
  pool: rpool
 state: ONLINE
  scan: scrub repaired 0 in 0h4m with 0 errors on Sun Sep 11 00:28:51 2016
config:

	NAME         STATE     READ WRITE CKSUM
	rpool*       ONLINE       0     0     0
	  /dev/sdd1* ONLINE       0     0     0

errors: No known data errors
```

**Comment by rlaager** [comment]
I'm marking this approved so my "Requested changes" is not blocking this.

As long as the default stays as is, this shouldn't be a problem. If the default were to change, then we'd have a big problem for GRUB and others that parse the output of `zpool status`. It might be worth considering that, even in the modified output, the * not be directly cuddled up next to the device name. That is, use `/dev/sda1 *` rather than `/dev/sda1*`.

One of the concerns above was about how to fit "ssd" in the `zpool iostat` output. One option would be to use "NR" instead, which could be written with N on the top line and the R on the bottom line. Then, you could have values of Y, N, and M (for mixed).

**Comment by tonyhutter** [comment]
**Crazy idea:**

Add a 'command' (-c) option to `zpool [status|iostat]` that runs an arbitrary command for each vdev and prints the result in the last column.  `$VDEV_PATH` is passed as an environment var for the command to use.  For example, this command would print out if the drive was rotational or not along with some other info:

```
$ zpool status -c 'lsblk --nodeps -no size,rota,model,name $VDEV_PATH'

	NAME        STATE     READ WRITE CKSUM  CMD
	mypool      ONLINE       0     0     0  
	  mirror-0  ONLINE       0     0     0  
	    A0      ONLINE       0     0     0  "7.3T    1       43000b40024812332"
	    A1      ONLINE       0     0     0  "7.3T    1       43000b4002482222a"
	    A2      ONLINE       0     0     0  "7.3T    1       43000b400248b2afa"
	    A3      ONLINE       0     0     0  "7.3T    1       43000b400248dcca2"
	    A4      ONLINE       0     0     0  "7.3T    1       43000b400248ba456"
	   sda      ONLINE       0     0     0  "7.3T    1 ST8000NM0075"
	   sdb      ONLINE       0     0     0  "7.3T    1 ST8000NM0075"
	   sdc      ONLINE       0     0     0  "120G    0 SV300S37A120G"
```
Another example:
```
$ zpool status -c 'smartctl -a $VDEV_PATH | grep "Current Drive Temperature:"'

	NAME        STATE     READ WRITE CKSUM  CMD
	mypool      ONLINE       0     0     0  
	  mirror-0  ONLINE       0     0     0  
	    A0      ONLINE       0     0     0  "Current Drive Temperature: 40C"
	    A1      ONLINE       0     0     0  "Current Drive Temperature: 32C"
	    A2      ONLINE       0     0     0  "Current Drive Temperature: 33C"
	    A3      ONLINE       0     0     0  "Current Drive Temperature: 43C"
	    A4      ONLINE       0     0     0  "Current Drive Temperature: 28C"
	   sda      ONLINE       0     0     0  "Current Drive Temperature: 20C"
	   sdb      ONLINE       0     0     0  "Current Drive Temperature: 20C"
	   sdc      FAULTED      0     0     0  "Current Drive Temperature: 109C"
```

Yes, it's a huge, giant hack, and not very elegant, but it's very extensible and would work on all platforms.  You would want `zpool` to run the commands in parallel for each vdev, of course.

**Comment by Rudd-O** [comment]
I would strongly prefer @tonyhutter 's suggestion instead.  It's more flexible, it can give out much better information, and it doesn't alter the default output of zpool, upon which many scripts already depend.

For showing whether a device is rotational or not, there could be a separate switch that adds a new column, or a separate utility that shows what ZFS thinks about the device.

**Comment by richardelling** [comment]
There is a thin line between crazy and clever, I think @tonyhutter has a clever idea here.
From a practical perspective, real monitoring is done separately and not scraped from zpool output. But for a quick glance, there are many times where it is convenient to run commands against the VDEV_PATH and join the output.

That said, actual implementation can be tricky. When a device is unhealthy, it can take a long time (minutes) to respond to any command. We do not want to hold any locks when this happens. Some reasonable timeout and feedback to the user for command failures is needed.

**Comment by behlendorf** [comment]
Upon seeing the options I actually think the dedicated column is probably best.  Even though I originally proposed otherwise.  It's going to be the simplest and most consistent way to do this.  Using `-S` seems reasonable to avoid a conflict with `zpool import` if this option gets extended to be available there.  Again, similar to `-PLg`.

**Comment by inkdot7** [comment]
Implemented with `-S`, now also in `zpool status`.  Example:
```
root@xcz:/home/htj/z/sep11/zfsid# cmd/zpool/zpool status -S
...
config:

	NAME       NR  STATE     READ WRITE CKSUM
	lcl         n  ONLINE       0     0     0
	  mirror-0  n  ONLINE       0     0     0
	    sdb3    n  ONLINE       0     0     0
	    sda3    n  ONLINE       0     0     0
	  mirror-1  y  ONLINE       0     0     0
	    sdd3    y  ONLINE       0     0     0
	    sdc3    y  ONLINE       0     0     0
	  mirror-2  m  ONLINE       0     0     0
	    sdb4    n  ONLINE       0     0     0
	    sdd4    y  ONLINE       0     0     0
```
Question: With the third commit (Set vdev_stat->vs_nonrotational...) I changed to set `vs_nonrotational` only once after `vdev_open`, instead of repeatedly in `vdev_get_stats_ex()`  Keeping the setting in `vdev_get_stats_ex()` would have the advantage of not getting out of sync with the `vdev_nonrot` for whatever reason (vdev add/remove/attach/detach).

Question: the boolean `nonrotational` had to be passed around a bit in `zpool_main.c` since `status_cbdata_t` is not passed to the printing routines like `iostat_cbdata_t` is.  Do an attempt of first passing that around, also taking `namewidth` and `name_flags` out of the argument lists?

Note: the two last commits (Do not force..., Repair indent...) are for unrelated issues found.

**Comment by richardelling** [comment]
A usability comment. I presume "NR" means "not-rotational"  However, specs like SCSI offer the following interpretation of the rotation value:
0 = not-reported
1 = ssd
>1 = HDD rotation (a best-effort value in the case of variable-rotation speed devices)
in many Linux drivers, this gets dumbed-down to be what is ultimately in the "rotational" value in sysfs, where 0 means "non-rotating" and 1 means "rotating"

In this UI, it seems that you're proposing a "non-rotating" column. So an SSD is reported as "not non-rotating" which in addition to being poor grammar is somewhat confusing to explain to the average person. It seems to me that pulling through the "rotational" value in sysfs is the best way to carry forward the notion that the device might be an HDD in the UI. A quick xref to the device's rotational setting in sysfs is very straightforward, even when the drivers are buggy and lie. 

In short, I propose the column as "rotational" or a shorthand of same. The values are boolean ("true"/"false" for humans or 0/1 for parseable).  Since vdevs do not have a corresponding rotational value in sysfs today, a value of "mixed" seems intuitive enough.

**Comment by tonyhutter** [comment]
> Question: the boolean nonrotational had to be passed around a bit in zpool_main.c since status_cbdata_t is not passed to the printing routines like iostat_cbdata_t is. Do an attempt of first passing that around, also taking namewidth and name_flags out of the argument lists?

I would love to see this cleaned up.  As you point out, it's annoying to pass in a million args to `print_status_config()` and friends.  Passing around single `status_cbdata_t` would be really nice.

Other notes:
- Is a virtual disk listed as rotational or non-rotational?  I'm thinking device mapper and multipath devices here. You could expand the definition of "NR" to a more general "disk type" column rather than just rotational/non-rotational.

- Could I request you use '-R' ('R'otational)' instead of '-S'?  I'm working on adding SMART data to `zpool status/iostat` at some point and wouldn't mind reserving -S for that.

- I see you're adding `vs_nonrotational` to `struct vdev_stat`.  Unfortunately you're going to run into issues if you have an older zfs.ko with a newer zpool binary, as zfs.ko is going to be reporting back the older version of the struct, which the newer `zpool` will misinterpret.  To get around it, you'll want to add "rotational" as an nvpair to ZPOOL_CONFIG_VDEV_STATS_EX (see `vdev_config_generate_stats()`).  Or alternately, just grab the sysfs value as @richardelling suggested, and you don't have to deal with `vdev_stat` at all.

**Comment by inkdot7** [comment]
I suppose @richardelling meant that an _HDD_ is reported as "not non-rotating", which indeed is confusing.  (The reason was that the exposed internal ZFS variable is named `vdev_nonrot`.)    If changed to `ROT` (and `rot` in `iostat`), then `-R` also makes more sense as flag.

A `vdev_disk` queries `blk_queue_nonrot`, while `vdev_file` sets `vdev_nonrot` to `B_TRUE`.  The general idea was to also expose what ZFS believes about a disk; querying sysfs in `zpool` would prevent that.

@tonyhutter thanks for the tip about `ZPOOL_CONFIG_VDEV_STATS_EX`, and I'll try with `status_cbdata_t`.

**Comment by inkdot7** [comment]
@tonyhutter when I tried about a year ago, I think that the linux dm did pass a non-rotational flag through for SSDs, but md did not.  (Whatever my memory is worth...)

Since then I've come to the conclusion that it makes little sense to glue SSDs together to match the size of a HDD in a mixed mirror.  Better to waste some HDD space and use several of them.

**Comment by inkdot7** [comment]
Updated to use `POOL_CONFIG_VDEV_STATS_EX`, and rebased on top of #5259 (first commit).

@tonyhutter
I did not change from `-S` to `-R` yet, as I noticed that `zpool import` uses `-R` for another purpose.  Since `zpool import` also can show a list of vdevs, using `-R` is perhaps not the best choice?

@richardelling 
I now used the heading `ROT` and then `yes` for spinning disks and `no` for solid-state disks.  However, although logical, now seeing it, I think this confusing too.  Since as a user, I would associate _yes_ with something good (like a solid-state drive), and _no_ with something not so good.
I would think that a heading of e.g. `SSD` and then yes and no for that would be better.  Or as suggested by @tonyhutter  earlier, a heading like `KIND` and then `ssd` and `hdd` instead of `yes` and `no`.  (`mix` stays `mix`).  In this latter case, one could also have `file` for vdevs which are files.  (Still easy to get out of the internal structures, so it continues to expose ZFS internal idea of things.)

(I'll be away from my test setup for some time, so it will be a while before I can test changes.)

For reference, the current look:

```
# cmd/zpool/zpool iostat -vS
                   capacity     operations     bandwidth 
pool        rot  alloc   free   read  write   read  write
----------  ---  -----  -----  -----  -----  -----  -----
lcl         mix  31,3G  1,99G      0      0  4,35K  6,19K
  sdd4      no   9,80G   269M      0      0  1,27K  1,06K
  sdc4      no   21,5G   765M      0      0  1,06K  1,52K
  mirror    mix   480K  1008M      0      0  2,02K  3,61K
    sdc3    no       -      -      0      0  1,04K  1,80K
    sda7    yes      -      -      0      0   1002  1,80K
----------  ---  -----  -----  -----  -----  -----  -----
```

```
# cmd/zpool/zpool status -S
  pool: lcl
 state: ONLINE
  scan: none requested
config:

	NAME       ROT  STATE     READ WRITE CKSUM
	lcl        mix  ONLINE       0     0     0
	  sdd4     no   ONLINE       0     0     0
	  sdc4     no   ONLINE       0     0     0
	  mirror-2 mix  ONLINE       0     0     0
	    sdc3   no   ONLINE       0     0     0
	    sda7   yes  ONLINE       0     0     0
```

**Comment by behlendorf** [comment]
> a heading like KIND and then ssd and hdd instead of yes and no. (mix stays mix).

This would allow us a lot more flexibility.  Not only would it allow types like **file** but I can imagine it being extended to cover other types like **zvol**, **loop**, **ram** which are easily identified by the major number.  Conceivably it could also be extended to identify **nvme** or **m2** devices.  But most importantly leaving this as generic as possible wouldn't rule out doing this kind of thing, even if we just start with `hdd`, `ssd`, `file`.

**Comment by tonyhutter** [comment]
I like the flexibility of a "type" or "kind" column as well.  "kind" also has the benefit that no one is using the `-k` or `-K` flags, so you could steal one of those if you wanted.

I would really advocate that you add it as a column to the end of the normal output.  The reason being that @behlendorf  and I were talking about how nice it would be if we could add additional columns at the end of `status|iostat` by using percentage signs to specify a column list.  Basically do it like the `date` command does.  For example:

```
%k:	Kind
%t:	Drive temperature
%u:	SMART uncorrectable errors
%l:	Drive lifetime

$ zpool status -c "%k%t%u%l"
  pool: lcl
 state: ONLINE
  scan: none requested
config:

    NAME       STATE     READ WRITE CKSUM  KIND  TEMP  UNCOR  LIFE
    lcl        ONLINE       0     0     0  
      sdd4     ONLINE       0     0     0   ssd   40C      0  333d
      sdc4     ONLINE       0     0     0   ssd   30C      0   50d
      mirror-2 ONLINE       0     0     0
        sdc3   ONLINE       2     5     0   hdd   97C     50   55d
        mpatha ONLINE       0     0     0  virt            
```
I'm not saying you need to implement all this.  Adding it as a flag the way you're doing it is fine.  I'm just asking to put "KIND" at the end, so that if we did do this, it wouldn't look awkward having KIND as the 1st column with the rest of the columns at the back.

**Comment by inkdot7** [comment]
@tonyhutter @behlendorf @richardelling   Changed to **kind**, used flag `-k` and moved to last column:

(Note that the entire pool is marked with `hdd`, as it has at least one top-level component that only is `hdd`)

```
# cmd/zpool/zpool iostat -vk
                    capacity     operations     bandwidth       
pool              alloc   free   read  write   read  write  kind
----------------  -----  -----  -----  -----  -----  -----  ----
lcl                406K  21,1G      5    114   402K  2,13M  hdd 
  sdd4            45,5K  10,1G      1     27  80,3K   382K  ssd 
  sdb3             188K  9,94G      1     26  80,3K   436K  hdd 
  mirror           172K  1008M      2     54   161K   956K  mix 
    sda3              -      -      1     26  80,3K   478K  hdd 
    sdd3              -      -      1     27  80,3K   478K  ssd 
  /root/zfsfile1      0    80M      2     12   157K   787K  file
----------------  -----  -----  -----  -----  -----  -----  ----

# cmd/zpool/zpool iostat -k
              capacity     operations     bandwidth       
pool        alloc   free   read  write   read  write  kind
----------  -----  -----  -----  -----  -----  -----  ----
lcl          406K  21,1G      4     87   307K  1,63M  hdd 

# cmd/zpool/zpool status -k
  pool: lcl
 state: ONLINE
  scan: none requested
config:

        NAME              STATE     READ WRITE CKSUM KIND
        lcl               ONLINE       0     0     0 hdd
          sdd4            ONLINE       0     0     0 ssd
          sdb3            ONLINE       0     0     0 hdd
          mirror-2        ONLINE       0     0     0 mix
            sda3          ONLINE       0     0     0 hdd
            sdd3          ONLINE       0     0     0 ssd
          /root/zfsfile1  ONLINE       0     0     0 file
```

File: man/man8/zpool.8 [hunk 1]
```diff
@@ -1571,6 +1571,15 @@ Scripted mode. Do not display headers, and separate fields by a single tab inste
 .sp
 .ne 2
 .na
+\fB\fB-k\fR\fR
+.ad
+.RS 12n
+Display the kind of device a pool or vdev is based on: ssd or hdd.  Mixed mirrored devices that have at least one solid-state member are marked "mix".  A pool is considered mixed if all members are at least mixed.
```
Comment (richardelling) [inline man/man8/zpool.8]: It is unclear from this description that a file is equivalent to solid-state.

Comment (inkdot7) [inline man/man8/zpool.8]: Thanks!  Added `A file vdev is assumed to have ssd characteristics when evaluating parents.`

Also changed the entries in the kind column to be right-aligned.

**Comment by kpande** [comment]
@tonyhutter why wouldn't it be `zpool status -o name,kind` ?

**Comment by tonyhutter** [comment]
@kpande `-o name,kind` works too. It's probably easier to type/remember as well.
@inkdot7  I'll take a look at your latest patch.

BTW - I ended up implementing my "crazy idea" -c command option (https://github.com/zfsonlinux/zfs/pull/5117#issuecomment-252118127). PR is https://github.com/zfsonlinux/zfs/pull/5368 if you want to take a look.

File: cmd/zpool/zpool_main.c [hunk 2]
```diff
@@ -2861,10 +2896,20 @@ print_iostat_header_impl(iostat_cbdata_t *cb, unsigned int force_column_width,
...
+		printf("      ");
+
+	printf("\n");
+
 	printf("%-*s", namewidth, title);
 
 	print_iostat_labels(cb, force_column_width, iostat_bottom_labels);
 
+	if (cb->cb_kind)
+		printf("  kind");
```
Comment (tonyhutter) [inline cmd/zpool/zpool_main.c]: The "kind" label should be printed in `print_iostat_labels()`

File: cmd/zpool/zpool_main.c [hunk 3]
```diff
@@ -1556,8 +1582,13 @@ print_status_config(zpool_handle_t *zhp, status_cbdata_t *cb, const char *name,
 		zfs_nicenum(vs->vs_write_errors, wbuf, sizeof (wbuf));
 		zfs_nicenum(vs->vs_checksum_errors, cbuf, sizeof (cbuf));
 		(void) printf(" %5s %5s %5s", rbuf, wbuf, cbuf);
+	} else {
+		(void) printf("                  ");
 	}
 
+	if (cb->cb_kind)
+		(void) printf(" %4s", kind_mark(nv));
```
Comment (inkdot7) [inline cmd/zpool/zpool_main.c]: In `status` the custom seemed to be just one space as separator.  The maximum entry is 4 bytes.

File: cmd/zpool/zpool_main.c (hunk 2 above)
Comment (inkdot7) [inline cmd/zpool/zpool_main.c]: That would seem to complicate things, sending some extra arguments, and ending up outside the loop in `print_iostat_labels()`.  Perhaps if `cb_kind` is turned into a `cb_flags`...  Should I look into that?

File: man/man8/zpool.8 [hunk 4]
```diff
@@ -2111,6 +2111,15 @@ Display vdev GUIDs instead of the normal device names. These GUIDs can be used i
 .sp
 .ne 2
 .na
+\fB\fB-k\fR\fR
+.ad
+.RS 12n
+Display the kind of device a pool or vdev is based on: ssd or hdd.  Mixed mirrored devices that have at least one solid-state member are marked "mix".  A pool is considered mixed if all members are at least mixed.  A file vdev is assumed to have ssd characteristics when evaluating parents.
```
Comment (behlendorf) [inline man/man8/zpool.8]: > Display the kind of device a pool or vdev is based on: ssd or hdd.

I'd suggest adding **file** to this list and any other valid kinds.

> A pool is considered mixed if all members are at least mixed.

This isn't clear to me.  What does _at least mixed_ mean?  I noticed in the example output  the pool itself was identified as **hdd** when I would have expected it to be **mixed**.

File: cmd/zpool/zpool_main.c [hunk 5]
```diff
@@ -1455,6 +1455,31 @@ max_width(zpool_handle_t *zhp, nvlist_t *nv, int depth, int max,
...
+	case VDEV_KIND_SSD:
+		return ("ssd");
+	case VDEV_KIND_FILE:
+		return ("file");
+	case VDEV_KIND_MIXED:
+		return ("mix");
+	case VDEV_KIND_HDD:
+		return ("hdd");
+	default:
+		return ("?");
```
Comment (behlendorf) [inline cmd/zpool/zpool_main.c]: What do you think about using either a `-` here or `unkn`?

File: cmd/zpool/zpool_main.c (hunk 2 above)
Comment (behlendorf) [inline cmd/zpool/zpool_main.c]: > Perhaps if cb_kind is turned into a cb_flag

That might be more consistent.  We're going to need to look in to the best way for these changes to co-exist with the proposed changes in #5368.

File: cmd/zpool/zpool_main.c (hunk 5 above)
Comment (inkdot7) [inline cmd/zpool/zpool_main.c]: `-` is easier on the eyes

**Comment by inkdot7** [comment]
@tonyhutter @behlendorf @richardelling I believe to have handled the outstanding issues.  Changing `cb_kind` to `cb_flag` turned out nicely.  Also reworded the man page entry; became smaller.  Did it become better?

**Comment by tonyhutter** [comment]
@inkdot7 apologies for not looking at this earlier.  I'm off for Christmas starting now and won't be back in the office until early January.  I can take a look at it then, or if @behlendorf approves it in the meantime, I'm fine with that too.

**Comment by tonyhutter** [comment]
Just gave this a test:
```
./cmd/zpool/zpool iostat -vk
                         capacity     operations     bandwidth       
pool                   alloc   free   read  write   read  write  kind
---------------------  -----  -----  -----  -----  -----  -----  ----
mypool                  336K  1008M      0      9  61.3K   252K   hdd
  mirror                336K  1008M      0      8  46.0K   204K   mix
    /tmp/inkdot/file1      -      -      0      3  15.3K  68.0K  file
    A0                     -      -      0      2  15.3K  68.0K   hdd
    A1                     -      -      0      2  15.3K  68.0K   hdd
logs                       -      -      -      -      -      -     -
  A2                       0  7.25T      0      0  15.3K  48.1K   hdd
cache                      -      -      -      -      -      -     -
  A3                     30K  7.28T      0      0  4.42K  1.76K   hdd
  A4                     10K  7.28T      0      0  4.43K  1.11K   hdd
---------------------  -----  -----  -----  -----  -----  -----  ----
```
You'll want to fix it so that `mypool` shows a `kind` of "mix" since it has a mix of disks and a file.  Or alternatively, just print the `kind` for the leaf vdevs, and print a dash (`-`) for the pool and parent vdevs.

Also, can you rebase this on top of master?

**Comment by inkdot7** [comment]
Thanks @tonyhutter for looking at this.  Rebased onto master.

It showing `hdd` for `mypool` is due to the `log` device being of `hdd` kind.  (`log` devices are apparently included in `vdev_open_children()`, while `cache` devices are not.)  As then some vdev (here `log`, but still) only has `hdd` performance, the entire pool is marked such.  Also, in practice it should make no difference, as one would only use ssd devices both for log and cache, and they do not on their own 'downgrade' the kind determined for the entire pool.  I think it makes sense this way?

**Comment by chrisrd** [comment]
Late to the party and completely bike shedding so feel free to ignore or disparage, but...

The "kind" column name feels quite foreign to me, could it be "type" instead? For what it's worth, both `zpool status` and `zpool iostat` have the `-t` flag available.

In the context of computing, "type" is something I'm used to seeing but I don't recall ever seeing "kind" used elsewhere - that's actually what caught my eye scanning the output example above, I did a bit of a double take on seeing "kind" and had to investigate further to see what it meant.

From a purely English language point of view, I would suggest a "type" is a more precise category versus a "kind" being a somewhat more broad, vague or fuzzy category.

**Comment by inkdot7** [comment]
@chrisrd , an easy change, but I'd need some consensus.
`-k` and kind was used as a result of the [comment on Oct 20](https://github.com/zfsonlinux/zfs/pull/5117#issuecomment-254968666).  Also, it was attempted to use a flag which is free also in `import` and `add` (which also can show device lists), see [comment on Oct 8](https://github.com/zfsonlinux/zfs/pull/5117#issuecomment-252421296).

**Comment by tonyhutter** [comment]
> As then some vdev (here log, but still) only has hdd performance, the entire pool is marked such. 

So it sounds to me like "kind" for non-leaf-vdevs means "the slowest drive in the group"?  Is that right?  If so, then `mirror` should be "hdd" instead of "mix", correct?  Just trying to understand the definition of "kind" for non-leaf-vdevs:
```
mypool                  336K  1008M      0      9  61.3K   252K   hdd
  mirror                336K  1008M      0      8  46.0K   204K   mix
    /tmp/inkdot/file1      -      -      0      3  15.3K  68.0K  file
    A0                     -      -      0      2  15.3K  68.0K   hdd
    A1                     -      -      0      2  15.3K  68.0K   hdd
...
```

Also, I'm fine with the column being called "type" but still using `-k`.

Can you also squash your commits?

**Comment by inkdot7** [comment]
Since a file dev is considered to have ssd performance, the mirror is marked `mix`:  Mix is for mirrors with hdds but at least one ssd, which thanks to #4334 give ssd-like performance when reading.  (Entire pools inherit mix too, if they have nothing worse in them.)  Perhaps "mix" could be changed to something better?  (But I do not have any suggestion.)  Squashed, but did not change kind yet, will do that when I'm at a computer where I can test.

**Comment by inkdot7** [comment]
Changed "kind" -> "type" when printed, kept `-k`.  For some consistency, also changed constants etc. in the code from `KIND` to `ROT_TYPE`.  Just `TYPE` was not good, as there are already many kinds of `TYPE`.  I am not sure if this really was an improvement.  I have the previous commit, if needed.

**Comment by chrisrd** [comment]
Thanks, my brain no longer does a double take on seeing `kind` :-)

More bike shedding: it's not really only about rotational devices any more (e.g. `file`) how about `MEDIA_TYPE` rather than `ROT_TYPE`?

**Comment by inkdot7** [comment]
Thanks, `MEDIA_TYPE` looks better in the code.  I also changed to print the headings as "media" and use the flag `-M`, which was free.  (But do have the previous commit printing "type" by `-k` handy).

**Comment by kpande** [comment]
imo we should just drop this patch and use tonyhutter's `-c` idea.

**Comment by richardelling** [comment]
@kpande agree, the -c option offers a large superset of this functionality

**Comment by chrisrd** [comment]
agree, `-c` is a superset, making this option unnecessary. The only thing missing from `-c` is a nicely formatted heading for the extra column. Perhaps a `-C string:  column heading for -c` option would be nice there? Sadly I don't have the time to try that myself.

**Comment by inkdot7** [comment]
What `-c` does not show is the inheritance for non-leaf-vdevs, nor does it expose the internal state of the zfs module, see [comment Oct 7](https://github.com/zfsonlinux/zfs/pull/5117#issuecomment-252105505).  That's the reason I continued adjusting this.

**Comment by kpande** [comment]
to me it's not a compelling enough reason to include this very specific code, when `-c` could be extended to give the information we need.

**Comment by tonyhutter** [comment]
I think the folks who are going to look at `MEDIA` fall into two categories:

1. People working on the various storage classes patches (like https://github.com/zfsonlinux/zfs/pull/4365 and https://github.com/zfsonlinux/zfs/issues/3779 ), who want to know what ZFS internally thinks the drive type is.  For those people this patch makes sense, since there's no other way to get at that internal data.


2. People who are interested in what type of drives are in the pool from an administrative standpoint.  For those people `-c` is a better option, since it requires no code changes, and is easier to extend to newer drive types as they come along (XPoint, XPoint DIMM, NVMe, multipath, etc).


Other random points:
- I just tested with two pools: all HDD behind multipath, and all SSD behind multipath, and was happy to see that the patch correctly identified the disks as HDD or SDD.
```
	NAME        STATE     READ WRITE CKSUM MEDIA
	jet1        ONLINE       0     0     0   ssd
	  mirror-0  ONLINE       0     0     0   ssd
	    U0      ONLINE       0     0     0   ssd
	    U1      ONLINE       0     0     0   ssd
	    L2      ONLINE       0     0     0   ssd
...
	NAME        STATE     READ WRITE CKSUM MEDIA
	mypool5     ONLINE       0     0     0   hdd
	  mirror-0  ONLINE       0     0     0   hdd
	    U10     ONLINE       0     0     0   hdd
	    U11     ONLINE       0     0     0   hdd
...
```


- I really think you should only print the `MEDIA` values for leaf vdevs. That's all that people are going to care about anyway. It would make for much simpler code, and remove any confusion over what "mixed" means.

- I do see down the road that we might change the definition of "ssd" to be something more specific.  For example, what is reported now as "ssd" now may be reported as "NVMe" in a later release, and that could break scripts.  I suppose you could add a line in the man page to say "these definitions may change in the future and should not be considered a stable API" or something like that.

**Comment by inkdot7** [comment]
> I do see down the road that we might change the definition of "ssd" to be something more specific. For example, what is reported now as "ssd" now may be reported as "NVMe" in a later release, and that could break scripts. I suppose you could add a line in the man page to say "these definitions may change in the future and should not be considered a stable API" or something like that.

Thanks @tonyhutter , added to the man page.

> I really think you should only print the MEDIA values for leaf vdevs. That's all that people are going to care about anyway. It would make for much simpler code, and remove any confusion over what "mixed" means.

Before removing that, I'd like to suggest that a mixed mirror vdev not is a common setup.  Admins that do such an setup would be likely to know what it means, perhaps even appreciate the confirmation.  Others will never see such output.  Unless they misconfigured, in which case I think they will be happy to have been alerted after figuring what it meant.

**Comment by behlendorf** [comment]
Closing.  This functionality is being added to `zpool iostat/status -c` in PR #6121.
